src/embedded_flash/helper.rs

```diff
@@ -0,0 +1,282 @@
+// Copyright 2019-2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// For compiling with std outside of tests.
+#![cfg_attr(feature = "std", allow(dead_code))]
+
+use core::iter::Iterator;
+use persistent_store::{StorageError, StorageResult};
+
+/// Reads a slice from a list of slices.
+///
+/// The returned slice contains the interval `[start, start+length)`.
+///
+/// # Preconditions
+///
+/// - The passed in slices must not overlap.
```

Comment:
Maybe I can just remove that comment, but the documentation doesn't precisely cover the `OutOfBounds`. Just wondered when looking at your implementation: What happens if our storage actually had 2 slices and we try to write to overlapping memory:

[0x0000..0x1000) [0x1000..0x2000)
                [0x0F00..0x1100)

Reply:
> the documentation doesn't precisely cover the OutOfBounds

I think `OutOfBounds` is correctly documented with:

```
# Errors

Returns `OutOfBounds` if the range is not within exactly one slice.
```

Because either it's outside all of them (usual meaning of out of bounds) or it's overlapping multiple slices (reusing the error name and factorizing the code, because users probably don't care, they don't expect errors).

> What happens if our storage actually had 2 slices and we try to write to overlapping memory

I'm not sure I see the link with this function. This function only reads from a slice of slices. There is no write.

Reply:
I was wondering how this eventuality is tested. But it's unrelated to the PR, and I changed the error documentation.